### PR TITLE
bpo-29469: Move constant folding at AST level

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue29469

### CODE REVIEW

**Comment by methane** [comment]
I'll check all comments in https://bugs.python.org/review/29469/
Please don't review for a while.

**Comment by rhettinger** [comment]
Overall, this looks great.  The code is remarkably clean.

**Comment by methane** [comment]
`Python/importlib.h` is conflicted very often.  So I'll fix it right before merge.

**Comment by serhiy-storchaka** [comment]
See my comment on the tracker: https://bugs.python.org/issue29469#msg297596

**Comment by serhiy-storchaka** [comment]
Could you please remove `asdl_ct.py` and simplify `ast_opt.c` by getting rid of useless functions like `astfold_boolop`? The AST types are rarely changed, and I think it would be not hard to update the cleaner code. If sometimes we will need the general code generator we could restore it from your patches.

File: Python/ast_opt.c [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+    case Str_kind:
+        return e->v.Str.s;
+    case Bytes_kind:
+        return e->v.Bytes.s;
+    case Ellipsis_kind:
+        return Py_Ellipsis;
+    case NameConstant_kind:
+        return e->v.NameConstant.value;
+    default:
+        assert(!is_const(e));
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: `Py_UNREACHABLE()` is now used in `compile.c`.

File: Python/ast_opt.c [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+        return NULL;
+    }
+}
+
+static int make_const(expr_ty node, PyObject *val, PyArena *arena)
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Move `static int` to a separate line. And similarly in other functions below.

File: Python/ast_opt.c [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+                COPY_NODE(node, arg);
+                return 1;
+            }
+        }
+        return 1;
+    }
+
+    typedef PyObject *(*unary_op)(PyObject*);
+    static const unary_op ops[] = {
+        0,
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: AFAIK in C99 you can write
```c
[Invert] = PyNumber_Invert,
[Not] = unary_not,
...
```

File: Python/ast_opt.c [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+        break;
+    default: // Unknown operator
+        return 1;
+    }
+
+    /* Avoid creating large constants. */
+    Py_ssize_t size = PyObject_Size(newval);
+    if (size == -1) {
+        if (PyErr_ExceptionMatches(PyExc_KeyboardInterrupt))
+            return 1;
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: `newval` leaked.

File: Python/ast_opt.c [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+        break;
+    case BitAnd:
+        newval = PyNumber_And(lv, rv);
+        break;
+    default: // Unknown operator
+        return 1;
+    }
+
+    /* Avoid creating large constants. */
+    Py_ssize_t size = PyObject_Size(newval);
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c:192]: `newval` can be NULL.

File: Python/ast_opt.c [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+        return NULL;
+    }
+}
+
+static int make_const(expr_ty node, PyObject *val, PyArena *arena)
+{
+    if (val == NULL) {
+        if(!PyErr_ExceptionMatches(PyExc_KeyboardInterrupt))
+            PyErr_Clear();
+        return 1;
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Shouldn't it return 0 in case of KeyboardInterrupt?

File: Python/ast_opt.c [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+    default:
+        assert(!is_const(e));
+        return NULL;
+    }
+}
+
+static int make_const(expr_ty node, PyObject *val, PyArena *arena)
+{
+    if (val == NULL) {
+        if(!PyErr_ExceptionMatches(PyExc_KeyboardInterrupt))
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Missed space after `if`. Missed braces.

File: Python/ast_opt.c [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+    expr_ty arg, idx;
+    slice_ty slice;
+
+    arg = node->v.Subscript.value;
+    slice = node->v.Subscript.slice;
+    if (node->v.Subscript.ctx != Load ||
+            !is_const(arg) ||
+            /* TODO: handle other types of slices */
+            slice->kind != Index_kind ||
+            !is_const(slice->v.Index.value)) {
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Move `{` to a new line.

File: Python/ast_opt.c [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+{
+    CALL(astfold_expr, expr_ty, node_->context_expr);
+    CALL_OPT(astfold_expr, expr_ty, node_->optional_vars);
+    return 1;
+}
+
+
+#undef CALL
+#undef CALL_OPT
+#undef CALL_SEQ
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]:
```c
#undef CALL_INT_SEQ
```

File: Python/ast_opt.c [hunk 10]
```diff
@@ -0,0 +1,619 @@
+/* File automatically generated by Parser/asdl_ct.py and ast_opt.ct. */
+#include "Python.h"
+#include "Python-ast.h"
+
+
+/* TODO: is_const and get_const_value is copied from Python/compile.c.
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: s/is/are/ ?

File: Lib/test/test_plistlib.py [hunk 11]
```diff
@@ -320,7 +320,6 @@ def test_tuple_members(self):
                     'second': [1, 2],
                     'third': [3, 4],
                 })
-                self.assertIsNot(pl2['first'], pl2['second'])
```
Comment (serhiy-storchaka) [inline Lib/test/test_plistlib.py]: Good catch! This test can fail in other implementations. It should be fixed in 3.6 too. Opened #4813 for this.

**Review by serhiy-storchaka — COMMENTED** [review]
It may be worth to document this change in What's New. Even without additional optimizations it can optimize more than the peephole optimizer.

File: Python/ast_opt.c [hunk 12]
```diff
@@ -176,11 +180,21 @@ static int fold_binop(expr_ty node, PyArena *arena)
...
+        PyErr_Clear();
+        return 1;
+    }
+
     /* Avoid creating large constants. */
     Py_ssize_t size = PyObject_Size(newval);
     if (size == -1) {
-        if (PyErr_ExceptionMatches(PyExc_KeyboardInterrupt))
-            return 1;
+        Py_DECREF(newval);
```
Comment (serhiy-storchaka) [inline Python/ast_opt.c]: Only before return.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM if fix a one simple error.

**Comment by serhiy-storchaka** [comment]
Generated files not up to date

**Comment by serhiy-storchaka** [comment]
Great! Many thanks @methane for your work!

Now I'll try to implement additional optimizations.

File: Python/ast_opt.c [hunk 13]
```diff
@@ -0,0 +1,649 @@
...
+    return PyBool_FromLong(!r);
+}
+
+static int
+fold_unaryop(expr_ty node, PyArena *arena)
+{
+    expr_ty arg = node->v.UnaryOp.operand;
+
+    if (!is_const(arg)) {
+        /* Fold not into comparison */
```
Comment (adrian17) [inline Python/ast_opt.c:83]: Looks like this makes [this](https://github.com/methane/cpython/blob/30edda97b4acd07da7a16c3de7726cd5a23148a9/Python/peephole.c#L348) peephole pass unnecessary too?

Comment (methane) [reply Python/ast_opt.c:83]: Ah, you're right.

Comment (methane) [reply Python/ast_opt.c:83]: Created #4863.  Thanks.

**Pull request merged**